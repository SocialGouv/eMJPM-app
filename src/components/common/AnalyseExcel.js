// Begining of analyse Excel: Not use for now

export const INVALID_CODE_POSTAL = ["75000"];
export const VALID_RESIDENCES = ["A Domicile", "En établissement"];
export const VALID_TYPES = [
  "Tutelle",
  "Curatelle",
  "Sauvegarde de justice",
  "Mesure ad hoc",
  "MAJ",
  "tutelle aux biens",
  "tutelle à la personne",
  "tutelle aux biens et à la personne",
  "curatelle simple aux biens",
  "curatelle simple à la personne",
  "curatelle simple aux biens et à la personne",
  "curatelle renforcée aux biens",
  "curatelle renforcée à la personne",
  "curatelle renforcée aux biens et à la personne",
  "sauvegarde de justice",
  "sauvegarde de justice avec mandat spécial"
];

export const VALID_CIVILITES = ["F", "H"];

const MANDATORY_COLUMNS = [
  "date_ouverture",
  "type",
  "code_postal",
  "ville",
  "civilite",
  "annee",
  "numero_dossier",
  "residence"
];

const REGEXP_CODE_POSTAL = /^(([0-8][0-9])|(9[0-5])|(2[AB]))[0-9]{3}$/;

export const isValidCodePostal = code_postal =>
  // force boolean
  !!(
    code_postal &&
    code_postal.toString().match(REGEXP_CODE_POSTAL) &&
    !INVALID_CODE_POSTAL.includes(code_postal.toString())
  );

// convert XLSX.utils.sheet_to_json data
export const cleanInputData = dataInput => {
  const cols = dataInput[0];
  const isValidColumn = col => MANDATORY_COLUMNS.includes(col);
  // convert Array of arrays into Array of Objects and filters only needed importable keys
  return dataInput
    .map(row =>
      row.reduce((a, c, i) => ({ ...a, ...(isValidColumn(cols[i]) ? { [cols[i]]: c } : {}) }), {})
    )
    .map(row => ({
      ...row,
      date_ouverture: new Date((row.date_ouverture - (25567 + 2)) * 86400 * 1000)
    }));
};

export const isValidResidence = residence =>
  residence && residence.toString().includes(VALID_RESIDENCES);

export const isValidAnnee = annee => annee && annee.toString().match(/^([12][0-9]){3}$/);
export const isValidType = type => type && type.toString().includes(VALID_TYPES);
export const isValidCivilite = civilite =>
  civilite && civilite.toString().includes(VALID_CIVILITES);

export const validateData = data => {
  const messages = {};
  const colums = Object.keys(data[0]);

  const absentColumns = MANDATORY_COLUMNS.filter(col => colums.indexOf(col) === -1);
  if (absentColumns.length) {
    messages.push(`Les colonnes suivants sont introuvables : ${absentColumns.join(", ")}`);
  }
  // group messages by row
  const rows = data.slice(1);
  rows.map((datum, i) => {
    const rowMessages = [];
    if (!isValidCodePostal(datum.code_postal)) {
      rowMessages.push(`Code Postal non valide : "${datum.code_postal}"`);
    }
    if (!isValidResidence(datum.residence)) {
      rowMessages.push(`Résidence non valide : "${datum.residence}"`);
    }
    if (!isValidAnnee(datum.annee)) {
      rowMessages.push(`Année non valide : "${datum.annee}"`);
    }
    if (!isValidType(datum.type)) {
      rowMessages.push(`Type de mesure non valide : "${datum.type}"`);
    }
    if (!isValidCivilite(datum.civilite)) {
      rowMessages.push(`Genre non valide : "${datum.civilite}"`);
    }
    if (rowMessages.length) messages[`Ligne ${i + 1}`] = rowMessages;
  });
  return messages;
};

export default validateData;
