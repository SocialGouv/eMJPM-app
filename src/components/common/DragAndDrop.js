import React from "react";
import apiFetch from "../communComponents/Api";

//
// render prop
// handle selection, onAdd, onRemove
// triggers props.onAdd and props.onRemove http method
// assume given objects have an "id" property
//
class DragAndDrop extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      imageURL: ""
    };

    this.handleUploadImage = this.handleUploadImage.bind(this);
  }

  handleUploadImage(ev) {
    ev.preventDefault();

    const data = new FormData();
    data.append("file", this.uploadInput.files[0]);

    apiFetch(`/mandataires/upload`, {
      method: "POST",
      body: data,
      headers: {
        "Content-Type": "multipart/form-data"
      }
    });
  }

  render() {
    return (
      <form onSubmit={this.handleUploadImage} enctype="multipart/form-data">
        <div>
          <input
            ref={ref => {
              this.uploadInput = ref;
            }}
            type="file"
            name="userFile"
          />
        </div>
        <div>
          <button>Upload</button>
        </div>
      </form>
    );
  }
}

export default DragAndDrop;
