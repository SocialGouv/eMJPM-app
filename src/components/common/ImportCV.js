import React from "react";
import apiFetch from "../communComponents/Api";

class ImportCV extends React.Component {
  handleUploadImage(ev) {
    ev.preventDefault();

    const data = new FormData();
    data.append("file", this.uploadInput.files[0]);

    apiFetch(`/mandataires/upload`, {
      method: "POST",
      body: data,
      headers: {
        "Content-Type": "multipart/form-data"
      }
    }).catch(e => {
      alert("Impossible de uploader le Curriculum Vitae");
      throw e;
    });
  }

  render() {
    return (
      <form onSubmit={this.handleUploadImage} >
        <div>
          <input
            ref={ref => {
              this.uploadInput = ref;
            }}
            type="file"
            name="userFile"
          />
        </div>
        <div>
          <button>Upload</button>
        </div>
      </form>
    );
  }
}

export default ImportCV;
