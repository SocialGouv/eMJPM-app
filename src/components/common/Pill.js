import styled from "styled-components";

export const Pill = styled.div`
  font-size: 18px;
  min-width: 80px;
  padding: 0 10px;
  height: 28px;
  line-height: 28px;
  border-radius: 2px;
  text-align: center;
  color: white;
  display: inline-block;
  margin-right: 10px;
  margin-top: 5px;
`;

export const PillDispo = ({ dispo, dispo_max }) => (
  <Pill
    style={{
      background: getColorFromDisponibilite(dispo_max - dispo)
    }}
  >
    {dispo || 0} / {dispo_max || 0}
  </Pill>
);

export const getColorFromDisponibilite = dispo => {
  if (dispo <= 0) {
    return "#f05659";
  } else if (dispo <= 5) {
    return "#eb9123";
  }
  return "#43b04a";
};

export default Pill;
