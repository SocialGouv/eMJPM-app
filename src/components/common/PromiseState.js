// render prop that trigger promise onClick and notifies promise resolution
class PromiseState extends React.Component {
  state = {
    status: "idle"
  };

  onClick = () => {
    this.setState(
      {
        status: "loading"
      },
      () => {
        this.props
          .onClick()
          .then(data => {
            this.setState({
              status: "success",
              data
            });
          })
          .catch(e => {
            this.setState({
              status: "error",
              data: e
            });
          });
      }
    );
  };
  render() {
    return <div onClick={this.onClick}>{this.props.render(this.state)}</div>;
  }
}

export default PromiseState;
