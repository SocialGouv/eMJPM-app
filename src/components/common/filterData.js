const sortByDispo = (a, b) => {
  const dispoA = parseFloat(a) || -Infinity;
  const dispoB = parseFloat(b) || -Infinity;
  if (dispoA < dispoB) {
    return -1;
  }
  if (dispoA > dispoB) {
    return 1;
  }
  return 0;
};

const sortMandataires = (a, b) =>
  sortByDispo(a.mesures_en_cours / a.dispo_max, b.mesures_en_cours / b.dispo_max);

const stringMatch = (str, needle) => {
  return str.toLowerCase().indexOf(needle.toLowerCase()) !== -1;
};

const filterData = (data, filters) => {
  let filteredMesures =
    data &&
    data.filter(datum => {
      return stringMatch(datum[filters.content], filters.filter);
    });
  return filteredMesures.sort(sortMandataires);
};

export default filterData;
