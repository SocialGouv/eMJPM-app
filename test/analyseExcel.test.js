import { cleanInputData, isValidCodePostal } from "../src/components/common/AnalyseExcel";

const VALID_CPS = ["13000", "89100", /*"97121", "97216",*/ "20000"];

describe("isValidCodePostal", () => {
  test("75000 should be invalid ", () => {
    expect(isValidCodePostal(75000)).toBe(false);
  });
  VALID_CPS.forEach(cp => {
    test(`${cp} should be valid`, () => {
      expect(isValidCodePostal(cp)).toBe(true);
    });
  });
});

// mock XLSX.utils.sheet_to_json
const jsonSheet = [
  [
    "date_ouverture",
    "type",
    null,
    "code_postal",
    "ville",
    "civilite",
    null,
    "annee",
    "numero_dossier",
    "residence",
    null,
    "adad",
    "mandataire_id"
  ],
  [
    43041,
    "tutelle",
    2,
    57100,
    "thionville",
    "n",
    null,
    4150,
    "RG20136Q",
    "à domicile",
    null,
    "q",
    10
  ],
  [
    42707,
    "curatelle",
    12,
    75005,
    "paris",
    "n",
    125,
    4000,
    "1236HPLY",
    "en établissement",
    null,
    "s",
    10
  ],
  [
    40395,
    "sauvegarde de justice ",
    52,
    85000,
    "marseille",
    "n",
    54,
    4500,
    "meourg120",
    "à domicile",
    null,
    "t",
    10
  ]
];

test("cleanInputData", () => {
  expect(cleanInputData(jsonSheet)).toMatchSnapshot();
});
