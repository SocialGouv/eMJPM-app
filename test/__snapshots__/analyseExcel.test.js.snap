// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanInputData 1`] = `
Array [
  Object {
    "annee": "annee",
    "civilite": "civilite",
    "code_postal": "code_postal",
    "date_ouverture": Date { NaN },
    "numero_dossier": "numero_dossier",
    "residence": "residence",
    "type": "type",
    "ville": "ville",
  },
  Object {
    "annee": 4150,
    "civilite": "n",
    "code_postal": 57100,
    "date_ouverture": 2017-11-02T00:00:00.000Z,
    "numero_dossier": "RG20136Q",
    "residence": "à domicile",
    "type": "tutelle",
    "ville": "thionville",
  },
  Object {
    "annee": 4000,
    "civilite": "n",
    "code_postal": 75005,
    "date_ouverture": 2016-12-03T00:00:00.000Z,
    "numero_dossier": "1236HPLY",
    "residence": "en établissement",
    "type": "curatelle",
    "ville": "paris",
  },
  Object {
    "annee": 4500,
    "civilite": "n",
    "code_postal": 85000,
    "date_ouverture": 2010-08-05T00:00:00.000Z,
    "numero_dossier": "meourg120",
    "residence": "à domicile",
    "type": "sauvegarde de justice ",
    "ville": "marseille",
  },
]
`;
